Allow redirecting subprocess stdout to our stderr etc.

ijackson opened the pull request:
Currently there is no safe or portable way to redirect a `Command`s stdout to the invoking program's stderr, nor vice versa.

There a comment in the source that suggests that this was intended, and trying to handle the fd shuffling problem properly, but it can never be reached other than by crazy unsafe stunts.

Instead, provide appropriate `From` impls.

<strike>I have only tested the implementation for Unix because I know how to do that.</strike>  There is now a Windows implementation which seems to pass the CI.  This branch currently contains a DO NOT MERGE commit to enable the Windows CI build on this MR branch.  That should be removed after code review and before merge.

Being trait impls, these are insta-stable.  The question is just whether we want to support platforms that cannot do this, and if so, whether we want that to be a compile error or are happy for it to be detected by `Command` methods returning errors.

Closes #79731

ijackson committed:
std::process: impl From<io::Stdout> (etc.) for Stdio

Signed-off-by: Ian Jackson <ijackson@chiark.greenend.org.uk>

rust-highfive commented:
r? @kennytm

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3486545875) [(plain)](https://github.com/rust-lang/rust/commit/d2c79d0cc16dbf9a252a19173361746f84c24196/checks/3486545875/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
    Checking object v0.26.2
    Checking hashbrown v0.11.0
    Checking std_detect v0.1.5 (/checkout/library/stdarch/crates/std_detect)
    Checking addr2line v0.16.0
error[E0277]: the trait bound `sys::windows::process::Stdio: From<io::stdio::Stdin>` is not satisfied
     |
     |
1425 |         Stdio::from_inner(inherit.into())
     |                           ^^^^^^^^^^^^^^ the trait `From<io::stdio::Stdin>` is not implemented for `sys::windows::process::Stdio`
     = help: the following implementations were found:
     = help: the following implementations were found:
               <sys::windows::process::Stdio as From<AnonPipe>>
               <sys::windows::process::Stdio as From<sys::windows::fs::File>>
     = note: required because of the requirements on the impl of `Into<sys::windows::process::Stdio>` for `io::stdio::Stdin`

error[E0277]: the trait bound `sys::windows::process::Stdio: From<io::stdio::Stdout>` is not satisfied
     |
     |
1454 |         Stdio::from_inner(inherit.into())
     |                           ^^^^^^^^^^^^^^ the trait `From<io::stdio::Stdout>` is not implemented for `sys::windows::process::Stdio`
     = help: the following implementations were found:
     = help: the following implementations were found:
               <sys::windows::process::Stdio as From<AnonPipe>>
               <sys::windows::process::Stdio as From<sys::windows::fs::File>>
     = note: required because of the requirements on the impl of `Into<sys::windows::process::Stdio>` for `io::stdio::Stdout`

error[E0277]: the trait bound `sys::windows::process::Stdio: From<io::stdio::Stderr>` is not satisfied
     |
     |
1483 |         Stdio::from_inner(inherit.into())
     |                           ^^^^^^^^^^^^^^ the trait `From<io::stdio::Stderr>` is not implemented for `sys::windows::process::Stdio`
     = help: the following implementations were found:
     = help: the following implementations were found:
               <sys::windows::process::Stdio as From<AnonPipe>>
               <sys::windows::process::Stdio as From<sys::windows::fs::File>>
     = note: required because of the requirements on the impl of `Into<sys::windows::process::Stdio>` for `io::stdio::Stderr`
For more information about this error, try `rustc --explain E0277`.
error: could not compile `std` due to 3 previous errors
Build completed unsuccessfully in 0:00:17
```

</details>

camelid commented on library/std/src/process.rs:
```diff
@@ -1400,6 +1400,90 @@ impl From<fs::File> for Stdio {
     }
 }
 
+#[stable(feature = "stdio_from", since = "1.20.0")]
```
nit: I think this should get its own feature gate and version number, even though it's insta-stable, to avoid confusing users.

ijackson committed:
std::process (windows): Implement From<io::Stdin> etc. for imp::Stdio

This involves a new variant `imp;::Stdio::InheritSpecific`.

Signed-off-by: Ian Jackson <ijackson@chiark.greenend.org.uk>

ijackson commented:
I have rebased onto current head, given it its own feature tag (and updated the version) and provided what I thiink is an implementation on Windows.  I don't have any Windows here so I may have to iterate through the CI, in which case sorry for the noise.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4143265272?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/5d0e28e2237e8073435ba87c093686018bf9ccbb/checks/4143265272/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/library/std/src/sys/unix/process/process_common.rs at line 398:
             Stdio::StaticFd(fd) => unsafe {
                 let duped = OwnedFd::from_raw_fd(cvt_r(|| libc::dup(fd))?);
                 Ok((ChildStdio::Owned(FileDesc::from_inner(duped)), None))
+            },
 
             Stdio::MakePipe => {
             Stdio::MakePipe => {
                 let (reader, writer) = pipe::anon_pipe()?;
Diff in /checkout/library/std/src/sys/windows/process.rs at line 364:
 impl Stdio {
 impl Stdio {
     fn to_handle(&self, stdio_id: c::DWORD, pipe: &mut Option<AnonPipe>) -> io::Result<Handle> {
-        let use_stdio_id = |stdio_id| {
-            match stdio::get_handle(stdio_id) {
-                Ok(io) => unsafe {
-                    let io = Handle::from_raw_handle(io);
-                    let ret = io.duplicate(0, true, c::DUPLICATE_SAME_ACCESS);
-                    io.into_raw_handle();
-                },
-                },
-                Err(..) => unsafe { Ok(Handle::from_raw_handle(c::INVALID_HANDLE_VALUE)) },
-            }
+        let use_stdio_id = |stdio_id| match stdio::get_handle(stdio_id) {
+            Ok(io) => unsafe {
+                let io = Handle::from_raw_handle(io);
+                let ret = io.duplicate(0, true, c::DUPLICATE_SAME_ACCESS);
+                io.into_raw_handle();
+            },
+            },
+            Err(..) => unsafe { Ok(Handle::from_raw_handle(c::INVALID_HANDLE_VALUE)) },
 
         match *self {
         match *self {
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/library/std/src/sys/unix/process/process_common/tests.rs" "/checkout/library/std/src/sys/unix/process/mod.rs" "/checkout/library/std/src/sys/unix/process/process_common.rs" "/checkout/library/std/src/sys/sgx/condvar.rs" "/checkout/library/std/src/sys/unix/os_str/tests.rs" "/checkout/library/std/src/sys/unix/process/process_vxworks.rs" "/checkout/library/std/src/sys/unix/process/process_unsupported.rs" "/checkout/library/std/src/sys/unix/android.rs"` failed.
Build completed unsuccessfully in 0:00:13
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4146217116?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/5d857dc56cc9b041729f4fea03015bb500066841/checks/4146217116/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test result: FAILED. 12268 passed; 1 failed; 110 ignored; 0 measured; 0 filtered out; finished in 136.60s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-12/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "12.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:13:36
```

</details>

ijackson commented:
The failing test is that for #30490.  The test case, and the code there, involves passing a static `0`, `1` or `2` to a function which says it _**consumes ownership of the specified file descriptor_, which is clearly UB.

I think I am going to have to break this MR up or work around this somehow.  Nnnng.

camelid commented:
Just FYI: `fixup!` commits are ***not*** automatically squashed upon merge.

ijackson commented:
> Just FYI: `fixup!` commits are _**not**_ automatically squashed upon merge.

Indeed.  This MR needs more work and an FCP, so there's no danger of them ending up in tree before I've had a chance to tell my git to autosquash them.  But thanks for the note.

ijackson committed:
std::process (unix): Implement From<io::Stdin> etc. for imp::Stdio

This involves adding a new variant `imp::Stdio::StaticFd`.

Signed-off-by: Ian Jackson <ijackson@chiark.greenend.org.uk>

ijackson commented:
Well, here's a version with Windows and which sidesteps whether `FileDesc` is allowed to contain 0/1/2.  Let's see if the CI likes it.  I have **not tested it on Windows at all**.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4183231895?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/ae5fd3de9b440c75ca75f7c384f16b7c0d5f359a/checks/4183231895/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/library/std/src/sys/unix/process/process_common.rs at line 407:
                     fd.duplicate()?
                 };
                 Ok((ChildStdio::Owned(fd), None))
+            }
 
             Stdio::MakePipe => {
             Stdio::MakePipe => {
                 let (reader, writer) = pipe::anon_pipe()?;
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/library/std/src/sys/unix/process/process_common/tests.rs" "/checkout/library/std/src/sys/unix/process/process_fuchsia.rs" "/checkout/library/std/src/sys/sgx/rwlock.rs" "/checkout/library/std/src/sys/unix/fd.rs" "/checkout/library/std/src/sys/sgx/net.rs" "/checkout/library/std/src/sys/unix/android.rs" "/checkout/library/std/src/sys/sgx/path.rs" "/checkout/library/std/src/sys/unix/process/process_common.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

ijackson commented:
Would someone be able to help with the Windows tests?  Ideally, by adapting my new doctests so they work on Windows.  (Breaking the Unix version while doing that would be fine; I can fix it up.)

Alternatively, if someone could suggest a command to run that will succeed and produce some nonempty stdlout output, and no stderr output, that will probably suffice and I can iterate it through the CI.

I am going to try setting the E-easy tag because I think this should be straightforward for someone familiar with Windows.

@rustbot modify labels +O-windows +E-help-wanted +E-easy

Mark-Simulacrum commented:
@ijackson Would you still be interested in driving this forward? My recommendation for Windows testing is to enable the relevant builder on PRs in src/ci/github-actions/ci.yml, and then run  ./x.py run src/tools/expand-yaml-anchors.

You can find the pr builder section by looking for `pr:` (should have the mingw-check, llvm-12, and tools builders), and adding the x86_64-msvc-1 builder to that. Hopefully that's enough that you can test this change successfully.

I'm going to assign myself as the reviewer for this PR, if you need further help please run `@rustbot ready` to get the labels updated so I have it in my queue to take a look.

@rustbot author

ijackson commented:
This has got to the top of my todo list again.  Sorry for the delay...

ijackson committed:
std::process (windows): refactor Stdio::to_handle slightly

We're going to want to reuse this bit of code.

ijackson committed:
DO NOT MERGE - RUN WINDOWS CI

My attempt at following the instructions in
  https://github.com/rust-lang/rust/pull/88561#issuecomment-1193392005

ijackson commented:
Gosh!  I wasn't expecting that to pass.  But I do see
```
2023-01-03T16:42:52.9693329Z [0m[0m[1m[32m   Doc-tests[0m std
2023-01-03T16:42:54.3561006Z 
2023-01-03T16:42:54.3561378Z running 1101 tests
...
2023-01-03T16:43:26.7132116Z test result: ok. 1085 passed; 0 failed; 16 ignored; 0 measured; 0 filtered out; finished in 32.36s
```
I am going to mark this ready now and request review even though I'm leaving the "DO NOT MERGE" Windows CI commit at the top.

ijackson commented:
@rustbot ready

ijackson commented:
@rustbot modify labels -E-easy -E-help-wanted

ijackson commented:
@rustbot labels -T-infra -O-windows

ijackson commented:
@rustbot labels -A-testsuite

Mark-Simulacrum commented:
r? libs-api for T-libs-api FCP

pitaj commented:
Ping from triage. Appears the merge conflicts are due to the DO NOT MERGE commit, so I'm ignoring that. I'm going to reroll for the FCP.

r? libs-api

dtolnay commented on library/std/src/process.rs:
```diff
@@ -1431,6 +1431,92 @@ impl From<fs::File> for Stdio {
 }
 
+#[stable(feature = "stdio_from_stdio", since = "1.58.0")]
+impl From<io::Stdin> for Stdio {
```
Questions about this impl &mdash; could you clarify whether this has any use that is not identical to `.stdin(Stdio::inherit())`?

I tried the following, but it didn't work as I'd naively expect:

```rust
use std::io::{self, Read};
use std::process::Command;

fn main() {
    let mut child = Command::new("whoami")
        .stdout(io::stdin())
        .spawn()
        .unwrap();

    let mut stdin = String::new();
    io::stdin().read_to_string(&mut stdin).unwrap();

    let status = child.wait().unwrap();
    assert!(status.success());

    println!("{:?}", stdin);
}
```

(I am on Linux.)

Instead of the child process's stdout feeding into the current process's stdin, the child's stdout goes to stdout and the current process stdin is unaffected.

The example code you gave in the documentation for this impl uses `.stdin(io::stdin())`. We would prefer that this always be written as `.stdin(Stdio::inherit())` though, correct?

dtolnay commented:
The other 2 impls (to support `.stdout(io::stderr())` and `.stderr(io::stdout())`) LGTM and I would be happy to propose FCP for the libs-api team after we've discussed whether stdin should be included here.

ijackson commented:
> Instead of the child process's stdout feeding into the current process's stdin, the child's stdout goes to stdout and the current process stdin is unaffected.

I think this is a bug in my implementation.  It also demonstrates that my tests are not sufficient.  I'm investigating.

ijackson committed:
squash! std::process (unix): Implement From<io::Stdin> etc. for imp::Stdio

Exclude stdin.

ijackson committed:
squash! std::process (windows): Implement From<io::Stdin> etc. for imp::Stdio

Exclude Stdin

ijackson committed:
fixup! std::process: impl From<io::Stdout> (etc.) for Stdio

ijackson committed:
fixup! squash! std::process (unix): Implement From<io::Stdin> etc. for imp::Stdio

ijackson committed:
std::process (unix): Update discussion of fd borrowing

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5589237432/jobs/10217209813) [(plain)](https://github.com/rust-lang/rust/commit/d202637aede3b35e570fc751e307d2ab68c7dbdc/checks/15137718381/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[group]Run git config --global core.autocrlf false
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
  lfs: false
  submodules: false
  set-safe-directory: true
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
##[group]Run src/ci/scripts/should-skip-this.sh
src/ci/scripts/should-skip-this.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
##[group]Run src/ci/scripts/verify-channel.sh
src/ci/scripts/verify-channel.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/collect-cpu-stats.sh
src/ci/scripts/collect-cpu-stats.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
##[group]Run src/ci/scripts/install-sccache.sh
src/ci/scripts/install-sccache.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/select-xcode.sh
src/ci/scripts/select-xcode.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/install-clang.sh
src/ci/scripts/install-clang.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/install-wix.sh
src/ci/scripts/install-wix.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/disable-git-crlf-conversion.sh
src/ci/scripts/disable-git-crlf-conversion.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
##[group]Run src/ci/scripts/install-msys2.sh
src/ci/scripts/install-msys2.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/install-mingw.sh
src/ci/scripts/install-mingw.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/install-ninja.sh
src/ci/scripts/install-ninja.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/enable-docker-ipv6.sh
src/ci/scripts/enable-docker-ipv6.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
##[group]Run src/ci/scripts/disable-git-crlf-conversion.sh
src/ci/scripts/disable-git-crlf-conversion.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/verify-line-endings.sh
src/ci/scripts/verify-line-endings.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
  CACHE_DOMAIN: ci-caches.rust-lang.org
  IMAGE: mingw-check
##[endgroup]
##[group]Run src/ci/scripts/verify-backported-commits.sh
src/ci/scripts/verify-backported-commits.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
##[group]Run src/ci/scripts/verify-stable-version-number.sh
src/ci/scripts/verify-stable-version-number.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
##[group]Run src/ci/scripts/run-build-from-ci.sh
src/ci/scripts/run-build-from-ci.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  PR_CI_JOB: 1
  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
  HEAD_SHA: d202637aede3b35e570fc751e307d2ab68c7dbdc
  SCCACHE_BUCKET: rust-lang-ci-sccache2
  TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
---
   Compiling yaml-merge-keys v0.4.1
   Compiling expand-yaml-anchors v0.1.0 (/checkout/src/tools/expand-yaml-anchors)
    Finished release [optimized] target(s) in 4.50s
##[endgroup]
error: .github/workflows/ci.yml is not up to date; please run `x.py run src/tools/expand-yaml-anchors`.
caused by: failed to parse src/ci/github-actions/ci.yml
caused by: while parsing node, found unknown anchor at line 333 column 17
##[group]Clock drift check
  local time: Tue Jul 18 15:14:38 UTC 2023
  network time: Tue, 18 Jul 2023 15:14:38 GMT
##[endgroup]
```

</details>

ijackson committed:
squash! DO NOT MERGE - RUN WINDOWS CI

ijackson commented:
@rustbot labels modify -A-testsuite -T-infra

This is tripping because of the DO NOT MERGE commit I'm using to have the CI run the new doctests on Windows.

ijackson commented:
FTAOD, the actual MR branch here is not in a fit state to merge - it needs some squashing etc.

So, assuming the FCP completes with approval, don't merge the branch right away.  I will tidy it and force push, first.

Lonami commented on library/std/src/process.rs:
```diff
@@ -1491,6 +1491,66 @@ impl From<fs::File> for Stdio {
+    ///
+    /// # fn test() -> Result<(), Box<dyn std::error::Error>> {
+    /// let output = Command::new("whoami")
+    // "whoami" is a command which exists on both Unix and Winodows,
```
```suggestion
    // "whoami" is a command which exists on both Unix and Windows,
```

teor2345 commented on library/std/src/process.rs:
```diff
@@ -1491,6 +1491,66 @@ impl From<fs::File> for Stdio {
     }
 }
 
+#[stable(feature = "stdio_from_stdio", since = "1.58.0")]
```
I think this magic constant is new since the PR was opened:
https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#updating-the-feature-gate-listing

```suggestion
#[stable(feature = "stdio_from_stdio", since = "CURRENT_RUSTC_VERSION")]
```

teor2345 replied:
(This applies to all instances of the compiler version.)

ijackson commented:
Thanks everyone for the approval, and the code review.  I will update this branch soon, so that we can merge it.

ijackson committed:
Fix typo

Co-authored-by: Lonami <totufals@hotmail.com>

ijackson committed:
Use CURRENT_RUSTC_VERSION in one stable attr

Co-authored-by: teor <teor@riseup.net>

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5785969299/job/15679710737) [(plain)](https://github.com/rust-lang/rust/commit/ff0d579fc4b99caf83755f03981d2e24a4df693d/checks/15679710737/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling hashbrown v0.14.0
   Compiling object v0.31.1
   Compiling miniz_oxide v0.7.1
   Compiling addr2line v0.20.0
error[E0711]: feature `stdio_from_stdio` is declared stable since 1.58.0, but was previously declared stable since 1.72.0-beta.1
     |
     |
1533 | #[stable(feature = "stdio_from_stdio", since = "1.58.0")]

For more information about this error, try `rustc --explain E0711`.
error: could not compile `std` (lib) due to previous error
Build completed unsuccessfully in 0:00:21
```

</details>

ijackson committed:
Use CURRENT_RUSTC_VERSION in the other attr

As per
  https://github.com/rust-lang/rust/pull/88561#pullrequestreview-1561942767

ijackson committed:
unix process: Use BorrowedFd::try_clone_to_owned

ijackson commented:
For convenience, I have made the squashed and tidied branch into a new MR: #114590

pitaj commented:
Opinions will obviously differ, but personally, what you've done feels less convenient.

dtolnay reviewed (approved):
Thank you! I have approved your tidied PR in https://github.com/rust-lang/rust/pull/114590#pullrequestreview-1616337899.
